Enable MMA for BF16 data types on Powerpc

shalinib-ibm opened the pull request:
Final Version

ggerganov reviewed (commented):
There are several stray printfs and debugging code. Also the indentation is not consistent. Seems like a work-in-progress - should be cleaned-up before consider merging.

shalinib-ibm commented:
> There are several stray printfs and debugging code. Also the indentation is not consistent. Seems like a work-in-progress - should be cleaned-up before consider merging.

Thank you @ggerganov . Moved it to draft. Will publish the final version soon.

shalinib-ibm committed:
Enable MMA for BF16 data type on ppc64le

This patch upstreams llamafile's cpu matrix multiplication
kernels for ppc64le using MMA builtins for BF16 data type.

This change results in 9x - 40x gains
in total speed S t/s (ie all tokens/total time), across
various batch sizes tested using llama-batched-bench benchmark.

The patch is tested with Meta-Lllama-3-8B,
and Mistral-7B models (BF16 models generated by using
llama-quantize from corresponding FP32 models) on an
IBM POWER10 machine.

Signed-off-by: Shalini Salomi Bodapati <Shalini.Salomi.Bodapati@ibm.com>
